manager/orchestrator/restart/restart.go

```diff
@@ -190,6 +189,87 @@ func (r *Supervisor) Restart(ctx context.Context, tx store.Tx, cluster *api.Clus
 	return nil
 }
 
+// TaskRestartDelay calculates and returns the next delay based t's RestartPolicy
+// Returns nil on error parsing the RestartPolicy
+func (r *Supervisor) TaskRestartDelay(ctx context.Context, t *api.Task) *time.Duration {
+	var restartDelay time.Duration
+	backoff := &api.BackoffPolicy{
+		Base:   defaults.Service.Task.Restart.Backoff.Base,
+		Factor: defaults.Service.Task.Restart.Backoff.Factor,
+		Max:    defaults.Service.Task.Restart.Backoff.Max,
+	}
+	if t.Spec.Restart != nil {
+		if t.Spec.Restart.Backoff != nil {
+			if a, b := gogotypes.DurationFromProto(t.Spec.Restart.Backoff.Base); b != nil && a >= 0 {
+				backoff.Base = t.Spec.Restart.Backoff.Base
+			}
+			if a, b := gogotypes.DurationFromProto(t.Spec.Restart.Backoff.Factor); b != nil && a >= 0 {
+				backoff.Factor = t.Spec.Restart.Backoff.Factor
+			}
+			if a, b := gogotypes.DurationFromProto(t.Spec.Restart.Backoff.Max); b != nil && a >= 0 {
+				backoff.Max = t.Spec.Restart.Backoff.Max
+			}
```

Comment:
i'm thinking about the hilarious case in which a user deliberately sets a large base time and a negative factor in order to restart faster and faster. i'm glad you're ensuring that that does not occur.

Reply:
on a more serious note, there's a lot of great error checking here, but I think you should also update the service validation logic in controlapi to stop a service with these invalid parameters from even being created. (that's in addition to these checks, not instead of.)